lib/config.js

```diff
@@ -212,127 +87,277 @@ class Config {
+            if (rootPath && !pathIsInside(path.dirname(localConfigFile), rootPath)) {
+                continue;
+            }
 
-        // Step 6: Merge in command line rules
-        if (this.options.rules) {
-            debug("Merging command line rules");
-            config = ConfigOps.merge(config, { rules: this.options.rules });
-        }
+            debug(`Loading ${localConfigFile}`);
+            const localConfig = ConfigFile.loadCached(localConfigFile);
 
-        // Step 7: Merge in command line globals
-        config = ConfigOps.merge(config, { globals: this.globals });
+            // Ignore empty config files
+            if (!localConfig) {
+                continue;
+            }
 
-        // Only override parser if it is passed explicitly through the command line or if it's not
-        // defined yet (because the final object will at least have the parser key)
-        if (this.parser || !config.parser) {
-            config = ConfigOps.merge(config, {
-                parser: this.parser
-            });
-        }
+            // Check for root flag
+            if (localConfig.root === true) {
+                rootPath = path.dirname(localConfigFile);
+            }
 
-        if (this.parserOptions) {
-            config = ConfigOps.merge(config, {
-                parserOptions: this.parserOptions
-            });
+            debug(`Using ${localConfigFile}`);
+            configs.push(localConfig);
+            searched.push(localConfigDirectory);
         }
 
-        // Step 8: Merge in command line plugins
-        if (this.options.plugins) {
-            debug("Merging command line plugins");
-            Plugins.loadAll(this.options.plugins);
-            config = ConfigOps.merge(config, { plugins: this.options.plugins });
-        }
+        if (!configs.length && !cache && !this.useSpecificConfig) {
 
-        // Step 9: Apply environments to the config if present
-        if (config.env) {
-            config = ConfigOps.applyEnvironments(config);
+            // Fall back on the personal config from ~/.eslintrc
+            debug("Using personal config file");
+            const personalConfig = this.getPersonalConfig();
+
+            if (personalConfig) {
+                configs.push(personalConfig);
+            } else if (!hasRules(this.options) && !this.options.baseConfig) {
+
+                // No config file, no manual configuration, and no rules, so error.
+                const noConfigError = new Error("No ESLint configuration found.");
+
+                noConfigError.messageTemplate = "no-config-found";
+                noConfigError.messageData = {
+                    directory,
+                    filesExamined: localConfigFiles
+                };
+
+                throw noConfigError;
+            }
         }
 
-        this.cache[directory] = config;
+        // Merged with any cached portion
+        configs.reverse();
+        cache = cache ? cache.concat(configs) : configs;
```

Comment:
Instead of doing this, would it be possible to append `cache` immediately after `configCache.getHierarchyLocalConfigs(localConfigDirectory)`? That way, `cache` wouldn't need to be referenced outside of the loop, and the function would just return `configs`.